### PR TITLE
pinot 0.9.3 (new formulae)

### PR DESCRIPTION
**Pull request by kbastani**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

These 4 formulae are the distributed system component and dependencies for installing Apache Pinot (https://pinot.apache.org/) as a homebrew service.

### CODE REVIEW

**Comment by SMillerDev** [comment]
These should all be separate reviews and use `service do` blocks instead of plist

File: Formula/pinot.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    <<~EOS
+      [ -z "$JAVA_OPTS" ] && export JAVA_OPTS="-Xms4G -Xmx8G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -Xloggc:gc-pinot-controller.log -Dlog4j2.configurationFile=/usr/local/etc/pinot/log4j2.xml"
+    EOS
+  end
+
+  def install
+    data = var/"lib"
+
+    inreplace "conf/pinot-controller.conf",
+      "controller.data.dir=/tmp/pinot/data/controller", "controller.data.dir=#{data}/pinot/data/controller"
```
Comment (carlocab) [inline Formula/pinot.rb:28]:
```suggestion
    inreplace "conf/pinot-controller.conf",
              "controller.data.dir=/tmp/pinot/data/controller",
              "controller.data.dir=#{data}/pinot/data/controller"
```
Align the arguments to `inreplace` if you have to split them up. Same below.

File: Formula/pinot.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      "pinot.zk.server", "controller.zk.str"
+
+    inreplace "conf/pinot-controller.conf",
+      "pinot.cluster.name", "controller.helix.cluster.name"
+
+    mv "conf", "pinot"
+    etc.install "pinot"
+    libexec.install_symlink etc/"pinot" => "conf"
+
+    (var+"lib/pinot/data/controller").mkpath
```
Comment (carlocab) [inline Formula/pinot.rb:40]:
```suggestion
    (var/"lib/pinot/data/controller").mkpath
```

File: Formula/pinotbase.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  depends_on "openjdk"
+
+  def install
+    libexec.install "lib"
+    libexec.install "plugins"
+
+    prefix.install "bin"
+    bin.env_script_all_files(libexec/"bin", Language::Java.java_home_env)
+    Dir["#{bin}/*.sh"].each { |f| mv f, f.to_s.gsub(/.sh$/, "") }
```
Comment (carlocab) [inline Formula/pinotbase.rb:20]:
```suggestion
    bin.glob("*.sh").each { |f| mv f, bin/f.basename(".sh") }
```
Or, at least escape the `.` in your regex.

File: Formula/pinotbroker.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
+class Pinotbroker < Formula
+  desc "Broker service component for Apache Pinot"
+  homepage "https://pinot.apache.org/"
+  url "https://downloads.apache.org/pinot/apache-pinot-0.9.3/apache-pinot-0.9.3-bin.tar.gz"
```
Comment (carlocab) [inline Formula/pinotbroker.rb:4]: Same comment as above re having the same source.

File: Formula/pinot.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          <true/>
+          <key>WorkingDirectory</key>
+          <string>#{var}</string>
+        </dict>
+      </plist>
+    EOS
+  end
+
+  test do
+    assert_match(/StartController/, shell_output("#{bin}/pinot-admin -h 2>&1"))
```
Comment (carlocab) [inline Formula/pinot.rb:80]:
```suggestion
    assert_match("StartController", shell_output("#{bin}/pinot-admin -h 2>&1"))
```
Same comment re the test. Though, you never install anything into `bin`, so I'm surprised this even does anything.

**Review by carlocab — COMMENTED** [review]
My comments are bit repetitive, since they apply to multiple formulae. If these really do deserve separate formulae, you should probably split them up into separate PRs to make reviews simpler.

But, broadly, this should:
1. Probably be a single formula instead of multiple different ones;
2. have better tests;
3. use `service` blocks;
4. remove `revision 1` lines;
5. avoid regexes in `assert_match` when not needed.

See below for slightly more detailed feedback.

**Comment by kbastani** [comment]
@carlocab That is excellent feedback. Thank you so much for the thorough review. All of your comments and revisions are correct. Pinot is a distributed system which means that it is uniquely designed to run as separate "microservices". It does have a single component recipe where all services are started together as a single JAR, which seems to be the solution here. For now, I'm going to close this PR and use your feedback to create a single formula with better tests. Thanks again!

**Comment by kbastani** [comment]
@carlocab Side note -- you should put your contact info or Twitter in your profile for companies looking to recruit good talent. :-)

**Comment by carlocab** [comment]
Thanks for the tip. I'm not too hard to find on Twitter, but I've just made it a little easier.

**Pull request closed without merging**